test/controllers/HistoryControllerTest.ts

```diff
@@ -41,29 +43,132 @@ export function HistoryControllerTest() {
+        };
+
+        const throwsAnError = () => {
+          if (!threwError) {
+            threwError = true;
+            throw new Error();
+          }
+        };
 
-      hashUtils.getValue.and.callFake(() => {
-        if (!threwError) {
-          threwError = true;
-          throw new Error();
-        }
+        historyController.hashUtils.getValue = jasmine.createSpy('getValue').and.callFake(throwsAnError);
+
+        $$(env.root).trigger(InitializationEvents.restoreHistoryState);
+
+        expect(historyController.hashUtils.getValue).toHaveBeenCalledTimes(_.size(env.queryStateModel.attributes));
       });
-      historyController = new HistoryController(env.root, Mock.mockWindow(), env.queryStateModel, env.queryController, hashUtils);
 
-      $$(env.root).trigger(InitializationEvents.restoreHistoryState);
+      describe('when logging analytics event', () => {
+        beforeEach(() => {
+          historyController = new HistoryController(env.root, {
+            model: env.queryStateModel,
+            queryController: env.queryController,
+            usageAnalytics: env.usageAnalytics,
+            window: window
+          });
+          historyController.hashUtils = fakeHashUtils;
+          $$(historyController.element).trigger(InitializationEvents.restoreHistoryState);
+        });
+
+        afterEach(() => {
+          window.location.hash = '';
+        });
+
+        const simulateHashModule = (key, value) => {
+          historyController.hashUtils.getValue = jasmine.createSpy('getValue').and.callFake((valueNeeded: string) => {
+            if (valueNeeded == key) {
+              return value;
+            }
+            return undefined;
+          });
+        };
+
+        it('should log an analytics event when q changes', () => {
+          simulateHashModule('q', 'foo');
+          window.dispatchEvent(new Event('hashchange'));
+          expect(historyController.usageAnalytics.logSearchEvent).toHaveBeenCalledWith(analyticsActionCauseList.searchboxSubmit, {});
+        });
 
-      expect(hashUtils.getValue).toHaveBeenCalledTimes(_.size(env.queryStateModel.attributes));
+        it('should log an analytics event when sort changes', () => {
+          simulateHashModule('sort', 'date ascending');
+          window.dispatchEvent(new Event('hashchange'));
+          expect(historyController.usageAnalytics.logSearchEvent).toHaveBeenCalledWith(analyticsActionCauseList.resultsSort, {
+            resultsSortBy: 'date ascending'
+          } as IAnalyticsResultsSortMeta);
+        });
+
+        it('should log an analytics event when a facet changes to select a value', () => {
+          historyController.model.registerNewAttribute('f:@foo', []);
+          simulateHashModule('f:@foo', ['bar']);
+          window.dispatchEvent(new Event('hashchange'));
+          expect(historyController.usageAnalytics.logSearchEvent).toHaveBeenCalledWith(analyticsActionCauseList.facetSelect, {
+            facetId: '@foo',
```

Comment:
I noticed that this test and all that follow expect the same result, so would it be better to store the result object in a variable `expectedValue` and reuse it rather than copying it across tests?

Maybe not because it will remove the result content out of the test? Not sure

Reply:
There could be an utility method that takes care of doing the assertion.